### PR TITLE
[4.4] Phan: Fix some instances of PhanTypeArraySuspicious

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
This is the result of the static code analyser phan. This fixes several issues which phan marks as PhanTypeArraySuspicious.

### Testing Instructions
Testable only through codereview.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: administrator/components/com_admin/src/Model/HelpModel.php [hunk 1]
```diff
@@ -128,7 +128,7 @@ public function getLangTag()
      */
     public function &getToc()
     {
-        if (!\is_null($this->toc)) {
+        if (\count($this->toc)) {
```
Comment (laoneo) [inline administrator/components/com_admin/src/Model/HelpModel.php:131]: This change here would mean that it reads the files as long as they array is empty and not only the first time.

**Comment by richard67** [comment]
Habe you checked that we safely can replace the initial null values with empty arrays? What if the null values were intentional and are checked somewhere else? In my opinion code review is not a sufficient test method, especially not if people only review the changes but not other places in the code where the changed code is used. I think it always needs real testing if something is broken or not, and not just code review without any further instructions.

File: administrator/components/com_admin/src/Model/HelpModel.php (hunk 1 above)
Comment (Hackwar) [reply administrator/components/com_admin/src/Model/HelpModel.php:131]: Yes, but further down it is definitely filled with some content. Thus this would be filled every time after this method has run.

**Comment by Hackwar** [comment]
Yes, I did check that the null value isn't used anywhere else.

**Comment by laoneo** [comment]
Hello @Hackwar, thank you for your pull request. It is a very good one.

This changes can break backwards compatibility, so to be on the safe side I wold earliest merge it into Joomla 5. Can you rebase it so people can start testing?

**Comment by Hackwar** [comment]
Since this PR does not break any backwards compatibility and instead fixes PHP notices, I'm not going to rebase this.

**Comment by laoneo** [comment]
Thank you very much for your answer, really appreciated. As you change some default values on variables in the Factory class, I fear this might have some side effects into the wild.

**Comment by Hackwar** [comment]
Yes, the default is changed, however the factory itself only accesses that as an array and without initialising it first, it would throw a notice. Any code currently accessing this would have to check on array AND if the array-entry is set anyway. There is no situation where this could create issues. The only thing I could think of would be people resetting the cache again by setting this back to null instead of an array and that would throw a notice. But it would do that in the current code too.

File: libraries/src/Factory.php [hunk 2]
```diff
@@ -55,7 +55,7 @@ abstract class Factory
      * @var    Cache
      * @since  1.7.0
      */
-    public static $cache = null;
+    public static $cache = [];
```
Comment (laoneo) [inline libraries/src/Factory.php]: Can you please revert this change or move the whole pr to 5.0? This property is public in a heavily used class, so it should not be changed in a minor release.

Comment (Hackwar) [reply libraries/src/Factory.php]: done

**Comment by Hackwar** [comment]
I changed the requested value.

File: libraries/src/Filter/InputFilter.php [hunk 3]
```diff
@@ -428,9 +428,9 @@ protected static function decodeFileData(array $data)
      */
     protected function decode($source)
     {
-        static $ttr;
+        static $ttr = [];
```
Comment (laoneo) [inline libraries/src/Filter/InputFilter.php:431]: Would this change not mean that it always runs trough the if below when `get_html_translation_table` returns an empty array? So the whole caching becomes then obsolete. But on the other side I'm not sure when `get_html_translation_table` returns an empty array.

File: libraries/src/Microdata/Microdata.php [hunk 4]
```diff
@@ -26,7 +26,7 @@ class Microdata
      * @var    array
      * @since  3.2
      */
-    protected static $types = null;
+    protected static $types = [];
```
Comment (laoneo) [inline libraries/src/Microdata/Microdata.php]: I know I'm a bit PITA here, but for me this would count as BC break, so I would do it on 5.0 to be on the safe side.

File: libraries/src/Filter/InputFilter.php (hunk 3 above)
Comment (Hackwar) [reply libraries/src/Filter/InputFilter.php:431]: When that function returns an empty array, we have vastly bigger problems, meaning that none of our output would be quoted/escaped. It is safe to assume that that array will never be empty.

File: libraries/src/Microdata/Microdata.php (hunk 4 above)
Comment (Hackwar) [reply libraries/src/Microdata/Microdata.php]: I reverted this. Not going to argue about this.

File: libraries/src/Filter/InputFilter.php (hunk 3 above)
Comment (laoneo) [reply libraries/src/Filter/InputFilter.php:431]: Ok, thanks.
